index.bs

```diff
@@ -4335,27 +4337,30 @@ partial interface MLGraphBuilder {
-        - If *a* is 1-dimensional, it is converted to a 2-dimensional tensor by prepending a 1 to its dimensions.
-        - If *b* is 1-dimensional, it is converted to a 2-dimensional tensor by by appending a 1 to its dimensions.
-        - If both *a* and *b* are 1-dimensional, the operation is a vector dot-product, which produces a scalar output.
 </div>
 
 <details open algorithm>
   <summary>
     To <dfn dfn-for=MLGraphBuilder>calculate matmul output sizes</dfn>, given |a| and |b| run the following steps:
   </summary>
   <div class=algorithm-steps>
-    1. Let |shapeA| be |a|'s [=MLOperand/shape=] and |sizeA| be |a|'s [=MLOperand/rank=].
-    1. Let |shapeB| be |b|'s [=MLOperand/shape=] and |sizeB| be |b|'s [=MLOperand/rank=].
-    1. If |sizeA| and |sizeB| is 1, return the [=/list=] « 1 ».
-    1. If |sizeA| is 1 and |sizeB| is not, then insert 1 in the front of |shapeA| to become [ 1 | |shapeA| ] and let |sizeA| be 2.
-    1. If |shapeA|[0] is not equal to |shapeB|[|sizeB| - 2], then [=exception/throw=] an "{{OperationError}}" {{DOMException}}.
-    1. If |sizeB| is 1 and |sizeA| is not, then append 1 to |shapeB| to become [ |shapeB| | 1 ] and let |sizeB| be 2.
-    1. If |shapeA|[|sizeA| - 1] is not equal to |shapeB|[0], then [=exception/throw=] an "{{OperationError}}" {{DOMException}}.
-    1. Let |shape| be an array whose size |size| is the maximum of |sizeA| and |sizeB|.
-    1. [=list/For each=] |index| in [=the range=] 0 to |size|, exclusive:
-        1. Set |shape|[|index|] to the maximum of |shapeA|[|index|] and |shapeB|[|index|].
-    1. Return |shape|.
+    1. Let |shapeA| be a [=list/clone=] of |a|'s [=MLOperand/shape=]
+    1. Let |sizeA| be the [=list/size=] of |shapeA|.
+    1. Let |shapeB| be a [=list/clone=] of |b|'s [=MLOperand/shape=]
+    1. Let |sizeB| be the [=list/size=] of |shapeB|.
+    1. If either |sizeA| or |sizeB| is less than 2, then [=exception/throw=] a "{{DataError}}" {{DOMException}}.
+    1. Let |colsA| be |shapeA|[|sizeA| - 1].
+    1. Let |rowsA| be |shapeA|[|sizeA| - 2].
+    1. Let |colsB| be |shapeB|[|sizeB| - 1].
+    1. Let |rowsB| be |shapeB|[|sizeB| - 2].
+    1. If |colsA| is not equal to |rowsB|, then [=exception/throw=] a "{{DataError}}" {{DOMException}}.
+    1. Let |slicedA| be a [=list/clone=] of |shapeA| with the last 2 items [=list/removed=].
+    1. Let |slicedB| be a [=list/clone=] of |shapeB| with the last 2 items [=list/removed=].
+    1. If both |sizeA| and |sizeB| are greater than 2, then let |outputShape| be the result of [=bidirectionally broadcasting the shapes=] |slicedA| and |slicedB|. If that returns failure, then [=exception/throw=] a "{{DataError}}" {{DOMException}}.
+    1. Otherwise, if both |sizeA| and |sizeB| are equal to 2, then let |outputShape| be an empty [=/list=].
+    1. Otherwise, let |outputShape| be the [=list/size|longer=] of |slicedA| or |slicedB|.
+    1. [=list/Append=] « |rowsA|, |colsB| » to |outputShape|.
```

Comment:
🤔 I don't think the explicit `if`/`else if`/`else` are needed here because broadcasting already works correctly either way, even with scalars and empty dimensions (it must). So if you strip off the trailing spatial dimensions and take just the leading batch dimensions of `A` and `B` for bidirectional broadcasting, then if `batchShapeA` and `batchShapeB` are empty, you will correctly get an empty `[]` back, and for any non-empty cases, you'll get the correct result too. So I think it would just be this:

```suggestion
    1. Let |batchShapeA| be a [=list/clone=] of |shapeA| with the spatial dimensions (last 2 items) [=list/removed=].
    1. Let |batchShapeB| be a [=list/clone=] of |shapeB| with the spatial dimensions (last 2 items) [=list/removed=].
    1. Let |outputShape| be the result of [=bidirectionally broadcasting the shapes=] |batchShapeA| and |batchShapeB|. If that returns failure, then [=exception/throw=] a "{{DataError}}" {{DOMException}}.
    1. [=list/Append=] « |rowsA|, |colsB| » to |outputShape|.
```

Right?